build-system/common/esbuild-babel.js

```diff
@@ -83,17 +83,51 @@ function getEsbuildBabelPlugin(
       build.onLoad({filter: /\.[cm]?js$/, namespace: ''}, async (file) => {
         const filename = file.path;
         const {contents, hash} = await batchedRead(filename, optionsHash);
-        const transformed = await transformContents(filename, contents, hash, {
-          ...babelOptions,
+
+        const transformed = await transformContents(
           filename,
-          filenameRelative: path.basename(filename),
-        });
+          contents,
+          hash,
+          getFileBabelOptions(babelOptions, filename)
+        );
         return {contents: transformed};
       });
     },
   };
 }
 
+const CJS_TRANSFORMS = new Set([
+  'transform-modules-commonjs',
+  'proposal-dynamic-import',
+  'syntax-dynamic-import',
+  'proposal-export-namespace-from',
+  'syntax-export-namespace-from',
+]);
+
+/**
+ * @param {!Object} babelOptions
+ * @param {string} filename
+ * @return {!Object}
+ */
+function getFileBabelOptions(babelOptions, filename) {
+  // Patch for leaving files within node_modules as esm, since esbuild will break when trying
+  // to process a module file that contains CJS exports. This function is called after
+  // babel.loadOptions, therefore all of the plugins from preset-env have already been applied.
+  // and must be disabled individually.
```

Comment:
nit: maybe move this paragraph to the function's JSDoc so that VSCode's intellisense can display it at call-sites?

Reply:
I prefer where it is. Hypothetically we could remove the patch and the function still would be ok to leave. The actual jsdoc should read closer to: `Updates the babel configuration with file-specific settings`